servers/zms/src/main/java/com/yahoo/athenz/zms/NotificationManager.java

```diff
@@ -0,0 +1,172 @@
+        userDomainPrefix = userDomain + ".";
+        if (notificationService != null) {
+            scheduledExecutor = Executors.newScheduledThreadPool(1);
+            scheduledExecutor.scheduleAtFixedRate(new PendingMembershipApprovalReminder(), 0, 1, TimeUnit.DAYS);
+        }
+    }
+
+    NotificationManager(final DBService dbService, final NotificationServiceFactory notificationServiceFactory) {
+        this.dbService = dbService;
+        this.notificationServiceFactory = notificationServiceFactory;
+        notificationService = this.notificationServiceFactory.create();
+        init();
+    }
+
+    void shutdown() {
+        if (scheduledExecutor != null) {
+            scheduledExecutor.shutdownNow();
+        }
+    }
+
+    void generateAndSendPostPutMembershipNotification(String domain, String org, Boolean auditEnabled, Boolean selfserve, Map<String, String> details) {
+        if (isNotificationFeatureAvailable()) {
+            Set<String> recipients = new HashSet<>();
+            if (auditEnabled == Boolean.TRUE) {
+                //get recipient role(s) from audit domain
+                Role domainRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org + "." + domain, false, true, false);
+                Role orgRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org, false, true, false);
+                if (domainRole != null) {
+                    recipients.addAll(domainRole.getRoleMembers().stream()
+                            .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+                }
+                if (orgRole != null) {
+                    recipients.addAll(orgRole.getRoleMembers().stream()
+                            .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+                }
+            } else if (selfserve == Boolean.TRUE) {
+                // get admin role from the request domain
+                Role adminRole = dbService.getRole(domain, "admin", false, true, false);
+                recipients.addAll(adminRole.getRoleMembers().stream()
+                        .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+            }
+            Notification notification = createNotification(NOTIFICATION_TYPE_MEMBERSHIP_APPROVAL, recipients, details);
+            notificationService.notify(notification);
+        }
+    }
+
+    Notification createNotification(String notificationType, Set<String> recipients, Map<String, String> details) {
+        String recDomain;
+        AthenzDomain domain;
+        Notification notification = new Notification(notificationType);
+        notification.setDetails(details);
+        for (String recipient : recipients) {
+            if (recipient.contains(":role.")) {
+                //recipient is of type role. Extract role members
+                recDomain = recipient.substring(0, recipient.indexOf(":"));
```

Comment:
there is no reason for us to call both recipient.contains and recipient.indexOf - first one is a wasted operation. we should something like
`
idx = recipient.indexOf(":role.");
if (idx != -1) {
    recDomain = recipient.substring(0, idx);
...
`